Src/FluentAssertions/Execution/AssertionScope.cs

```diff
@@ -78,11 +83,36 @@ public AssertionScope(string context)
             Context = context;
         }
 
+        internal Lazy<string> ContextPromise { get; set; }
+
+        internal void ResolveContext()
+        {
+            if (ContextPromise != null)
+            {
+                context = ContextPromise.Value;
+                ContextPromise = null;
+            }
+        }
+
         /// <summary>
         /// Gets or sets the context of the current assertion scope, e.g. the path of the object graph
         /// that is being asserted on.
         /// </summary>
-        public string Context { get; set; }
+        public string Context
+        {
+            get
+            {
+                ResolveContext();
+
+                return context;
+            }
+
+            set
```

Comment:
🤔 I think having `Context` and `ContextPromise` at the same time is confusing. Maybe we should change the `Context` property take a delegate, just like `GetSubjectId`. What do you think @jnyrup ?

Reply:
Just to get this right, are we invoking the `GetSubjectId` delegate when not failing?
My hope is that we *only* call it when failing.

But I will have better time to look in this tomorrow.

Reply:
> My hope is that we only call it when failing.

That's the idea, yes.

Reply:
Hmm, well, perhaps it's a peculiarity of my new equivalency steps and not generally the case, but I had thought I'd copied an existing pattern. When I profiled it the first time, it seemed to be taking a stack trace and then scanning the associated source code file over and over with no test failures. It seems to happen when recursively calling `parent.AssertEqualityUsing` -- that calls `UpdateScopeWithReportableContext` which in turn reads `CurrentNode.Description`.

Reply:
Hmm, so... If I'm understanding correctly, `Context` is only actually _read_ during failure scenarios when the `expectation` delegate is called and the `Continuation` returned by `FailWith` is resolved? So, it would work to replace `Context` with a delegate like `GetSubjectId`, or perhaps a `Lazy<string>`, _except_ for one wrinkle: When I straight-up deferred the calculation, two tests started failing: Tests that override the handling for a property with a `.Using` that _itself_ starts a new assertion would previously have calculated the `Context` early on and latched onto the original line of code, but with deferred processing, it'd find the nested call inside the `.Using` instead. When trying to "magically" inject the expression on which the assertion failed, it started injecting the inner one instead, and fortunately there are tests that explicitly check for that :-)
```
            // Act
            Action act = () => subject.Should().BeEquivalentTo(other,
                o => o
                    .Using<string>(c => c.Subject.Should().Be(c.Expectation))
                    .When(si => si.Path == "Id"));

            // Assert
            act.Should().Throw<XunitException>()
                .Which.Message.Should()
                .Contain("Expected property subject.Id (of type double?) to be <null>, but found \"foo\"")
                    .And.NotContain("from expectation");
```
If `Context` calculation is deferred as late as possible, then by the time it's calculating it, the `c => c.Subject.Should().Be(c.Expectation)` is on the call stack before the `subject.Should().BeEquivalentTo`, and so the actual message says `"Expected property c.Subject.Id...` instead of `"Expected property subject.Id...`.

That's why there's a `ResolveContext` method that gets called by `AssertionRuleEquivalencyStep.cs` right before it recurses into the `.Using`.

Reply:
The changes as written don't break the API surface, because the public `Context` property is still of type `string`. If it's changed to a `Lazy<string>`, that's a breaking change. But... that's okay because a new major version is about to be released??

Reply:
I don't think we would mind breaking this. It's quite unusual for somebody to use that API.